### PR TITLE
Don't render annotations with a null dimension

### PR DESCRIPTION
**Pull request by calixteman**
Fixes issue #17906.

### CODE REVIEW

File: test/unit/display_utils_spec.js [hunk 1]
```diff
@@ -35,16 +35,22 @@ describe("display_utils", function () {
...
 
-    it("`create` should throw an error if the dimensions are invalid", function () {
-      // Invalid width.
-      expect(function () {
-        return canvasFactory.create(-1, 1);
-      }).toThrow(new Error("Invalid canvas size"));
+    it("`create` should reset dimensions if they're invalid", function () {
+      if (isNodeJS) {
+        pending("Document is not supported in Node.js.");
+      }
```
Comment (Snuffleupagus) [inline test/unit/display_utils_spec.js:41]: Please don't just disable testing in Node.js environments, but instead update the tests to still work there!

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]

File: test/unit/display_utils_spec.js (hunk 1 above)
Comment (timvandermeij) [reply test/unit/display_utils_spec.js:41]: Unless I'm mistaken here, is that possible here here looking at the test below it where this is also done at https://github.com/mozilla/pdf.js/pull/17908/files#diff-2b17e0587d29eb901056a2a81f40706775550a117b3d1fa7cd2a2a29e1083144R57-R59? Previously we only threw an exception which was testable in all environments, but now we're actually creating a canvas element and that requires `document` to exist. If I remove the `if` statement for the existing test, which also creates a canvas element, it fails with `TypeError: Cannot read properties of undefined (reading 'createElement'` which is probably why the existing tests are already done like this.

If there is a way in which this can be done though, that'd be great because then we can also enable the already-existing tests in Node.js I think. Unless it's easy I'm not sure it should be in scope of this PR though.

Comment (Snuffleupagus) [reply test/unit/display_utils_spec.js:41]: In this case the "proper" solution, but probably best for a follow-up, would be to add a new `node_utils_spec.js` test file (only run in Node.js) with `NodeCanvasFactory` tests that are essentially a copy of these ones (modified as needed):
https://github.com/mozilla/pdf.js/blob/30b65ca0678173af8fad4cf99cc42f0d5ec2425d/test/unit/display_utils_spec.js#L27-L117

File: src/display/base_factory.js [hunk 2]
```diff
@@ -45,9 +45,8 @@ class BaseCanvasFactory {
   }
 
   create(width, height) {
-    if (width <= 0 || height <= 0) {
-      throw new Error("Invalid canvas size");
-    }
+    width = Math.max(width, 0);
+    height = Math.max(height, 0);
```
Comment (Snuffleupagus) [inline src/display/base_factory.js:49]: It seems quite unfortunate to modify a general utility such as `CanvasFactory` just to fix one annotation bug.

Could we perhaps instead return an *empty* `operatorList`, from the src/core/annotation.js code, when the annotation has "bad" dimensions?

Comment (timvandermeij) [reply src/display/base_factory.js:49]: I have been considering that option too and went a bit back and forth on where the responsibility of this should be in the code. In the end I also went for the idea of changing the `CanvasFactory` because I didn't really see a reason to disallow e.g. creating 0 x 0 canvases (even though it's probably not very useful to have such canvases, but breaking on it also seemed a bit unnecessary) and it might fix a whole class of similar bugs if we fix this in the factory itself as it's used in various places. However, I can also see the benefit of solving this locally for annotations only because it makes the scope smaller and that's the specific issue we have seen here.

In short, I don't really have a clear preference for either solution and think both could be fine here.

Comment (Snuffleupagus) [reply src/display/base_factory.js:49]: To me it makes no sense to create a `0 x 0` canvas, since that basically points to a bug elsewhere, and these changes also make the `create` and `reset` methods *inconsistent* w.r.t. dimension validation which seems quite unfortunate.

Given that we ignore empty annotations in the display layer, [see here](https://github.com/mozilla/pdf.js/blob/a208d6bca72f8cb700f7bb506ea1bc3e76cbaec0/src/display/annotation_layer.js#L2986-L2989), it seems reasonable to me to generate *empty* operatorLists for such cases as well.

Comment (timvandermeij) [reply src/display/base_factory.js:49]: That's a good point. I didn't know we already ignored empty annotations in that code already, so in that case the empty operator list solution indeed seems overall more consistent here.

**Review by Snuffleupagus — APPROVED** [review]
